include/smeagle/parameter.h

```diff
@@ -16,13 +17,47 @@ namespace smeagle {
    * an actual parameter at a callsite, or a return value from a function.
    *
    */
-  struct parameter {
-    std::string name;
-    std::string type;
-    std::string direction;
-    int pointer_indirections;
-    std::string location;
-    size_t size_in_bytes;
+  class parameter {
+  public:
+    template <typename T> explicit parameter(T x) : self(std::make_shared<model<T>>(std::move(x))) {}
+
+    std::string name() const { return self->name(); }
```

Comment:
Oh wow! So is this C++ very fugly way of doing something like class inheritance? E.g., we can now create different parameter types that expose these functions?

Reply:
This is actually the opposite of class inheritance. It's called the Non-Virtual Interface (NVI). Instead of saying "I am a base class that exposes function 'foo' to be overridden by child types", it says "I have this member function, and I will accept any type that also has this member function". It allows for polymorphism, but without creating a hierarchy of types. It's really more like Python's duck typing (gurus would call it contract-based design).